src/passes/staticArrayIndexer.ts

```diff
@@ -0,0 +1,207 @@
+import { generateExpressionTypeString } from '../utils/getTypeString';
+import { createIdentifier } from '../utils/nodeTemplates';
+import { specializeType } from '../utils/nodeTypeProcessing';
+import { typeNameFromTypeNode } from '../utils/utils';
+
+export class StaticArrayIndexer extends ASTMapper {
+  /*
+    This pass change replace all non literal calldata static array index access for 
+    non literal MEMORY static array index access.
+    This pass is needed because static arrays are handled as cairo tuples, but cairo
+    tuples (for now) can only be indexed by literals
+    e.g.
+      uint8[3] calldata b = ...
+      uint8 x = b[i]
+   // gets replaced by:
+      uint8[3] calldata b = ...
+      uint8[3] memory mem_b  = b
+      uint8[x] = mem_b[i]
+  */
+
+  // Tracks calldata arrays which already have a memory counterpart
+  // initalized
+  private staticArrayAccesed = new Map<number, VariableDeclaration>();
+  visitMemberAccess(node: MemberAccess, ast: AST): void {
+    this.staticIndexToMemory(node, ast);
+  }
+
+  visitIndexAccess(node: IndexAccess, ast: AST): void {
+    this.staticIndexToMemory(node, ast);
+  }
+
+  private staticIndexToMemory(node: IndexAccess | MemberAccess, ast: AST): void {
+    if (!hasIndexAccess(node, ast)) {
+      return this.visitExpression(node, ast);
+    }
+
+    const identifier = getReferenceDeclaration(node, ast);
+
+    assert(identifier.vReferencedDeclaration instanceof VariableDeclaration);
+    const parentFunction = node.getClosestParentByType(FunctionDefinition);
+    assert(
+      parentFunction !== undefined,
+      'Calldata types must be enclosed in a function definition',
+    );
+
+    const refId = identifier.referencedDeclaration;
+
+    let refVarDecl = this.staticArrayAccesed.get(refId);
+    if (refVarDecl === undefined) {
+      refVarDecl = this.initMemoryArray(identifier, parentFunction, ast);
+      this.staticArrayAccesed.set(refId, refVarDecl);
+    }
+
+    const indexReplacement = this.setExpressionToMemory(node, refVarDecl, ast);
+    ast.context.register(node);
+    ast.replaceNode(node, indexReplacement);
+  }
+
+  private setExpressionToMemory(
+    expr: Expression,
+    refVarDecl: VariableDeclaration,
+    ast: AST,
+  ): Identifier | IndexAccess | MemberAccess {
+    const exprType = getNodeType(expr, ast.compilerVersion);
+    const exprMemoryType = specializeType(generalizeType(exprType)[0], DataLocation.Memory);
+
+    let replacement: Identifier | IndexAccess | MemberAccess;
+    if (expr instanceof Identifier) {
+      replacement = createIdentifier(refVarDecl, ast, DataLocation.Memory);
+    } else if (expr instanceof IndexAccess) {
+      assert(expr.vIndexExpression instanceof Expression);
+      replacement = new IndexAccess(
+        expr.id,
+        expr.src,
+        generateExpressionTypeString(exprMemoryType),
+        this.setExpressionToMemory(expr.vBaseExpression, refVarDecl, ast),
+        cloneASTNode(expr.vIndexExpression, ast),
```

Comment:
Is this clone necessary? I would have thought that if the old node is getting replaced then it should suffice to move its children without cloning them